coins: warn on oversized `-dbcache`

l0rinc opened the pull request:
### Summary

Oversized allocations can cause out-of-memory errors or [heavy swapping](https://github.com/getumbrel/umbrel-os/issues/64#issuecomment-663637321), [grinding the system to a halt](https://x.com/murchandamus/status/1964432335849607224).

### Fix

Added a minimal system helper to query total physical RAM on [Linux/macOS/Windows](https://stackoverflow.com/a/2513561) (on unsupported platforms we just disable this warning completely).
The added test checks if the value is roughly correct by checking if the CI platforms are returning any value and if the value is at least 1 GB (as a simple property test checking if the unit size is correct, e.g. doesn't return megabytes or bits).

### Details

`LogOversizedDbCache()` now emits a startup warning if the configured `-dbcache` exceeds a cap derived from system RAM, using the same parsing/clamping as cache sizing via `CalculateDbCacheBytes()`. This isn't meant as a recommended setting, rather a likely upper limit.

Note that we're not modifying the set value, just issuing a warning.
Also note that the 75% calculation is rounded for the last two numbers since we have to divide first before multiplying, otherwise we wouldn't stay inside `size_t` on 32-bit systems - and this was simpler than casting back and forth.

We could have chosen the remaining free memory for the warning (e.g. warn if free memory is less than 1 GiB), but this is just a heuristic, we assumed that on systems with a lot of memory, other processes are also running, while memory constrained ones run only Core.

### Cap

If total RAM < 2 GiB, cap is `DEFAULT_DB_CACHE` (`450 MiB`), otherwise it's 75% of total RAM.
The threshold is chosen to be close to values commonly used in [raspiblitz](https://github.com/raspiblitz/raspiblitz/blob/dev/home.admin/_provision.setup.sh#L98-L115) for common setups:

| Total RAM | `dbcache` (MiB) | raspiblitz % | proposed cap (MiB) |
|----------:|----------------:|-------------:|-------------------:|
|     1 GiB |             512 |        50.0% |               450* |
|     2 GiB |            1536 |        75.0% |               1536 |
|     4 GiB |            2560 |        62.5% |               3072 |
|     8 GiB |            4096 |        50.0% |               6144 |
|    16 GiB |            4096 |        25.0% |              12288 |
|    32 GiB |            4096 |        12.5% |              24576 |

[Umbrel issues](https://github.com/getumbrel/umbrel-os/issues/64#issuecomment-663816367) also mention 75% being the upper limit.

### Reproducer

Starting `bitcoind` on an 8 GiB rpi4b with a dbcache of 7 GiB:
> ./build/bin/bitcoind -dbcache=7000

warns now as follows:
```
2025-09-07T17:24:29Z [warning] A 7000 MiB dbcache may be too large for a system memory of only 7800 MiB.
Warning: A 7000 MiB dbcache may be too large for a system memory of only 7800 MiB.
2025-09-07T17:24:29Z Cache configuration:
2025-09-07T17:24:29Z * Using 2.0 MiB for block index database
2025-09-07T17:24:29Z * Using 8.0 MiB for chain state database
2025-09-07T17:24:29Z * Using 6990.0 MiB for in-memory UTXO set (plus up to 286.1 MiB of unused mempool space)
```

### Manual testing

Besides the [godbolt](https://godbolt.org/z/ec81Tjvrj) reproducers for the new total memory method, we also tested the warnings manually on:
- [x] Apple M4 Max, macOS 15.6.1
- [x] Intel Core i9-9900K, Ubuntu 24.04.2 LTS
- [x] Raspberry Pi 4 Model B, Armbian Linux 6.12.22-current-bcm2711
- [x] Intel Xeon x64, Windows 11 Home Version 24H2, OS Build 26100.4351

l0rinc committed:
coins: warn on oversized -dbcache

Oversized allocations can cause out-of-memory errors or [heavy swapping](https://github.com/getumbrel/umbrel-os/issues/64#issuecomment-663637321), [grinding the system to a halt](https://x.com/murchandamus/status/1964432335849607224).

`LogOversizedDbCache()` now emits a startup warning if the configured `-dbcache` exceeds a cap derived from system RAM, using the same parsing/clamping as cache sizing via CalculateDbCacheBytes(). This isn't meant as a recommended setting, rather a likely upper limit.

Note that we're not modifying the set value, just issuing a warning.
Also note that the 75% calculation is rounded for the last two numbers since we have to divide first before multiplying, otherwise we wouldn't stay inside size_t on 32-bit systems - and this was simpler than casting back and forth.

We could have chosen the remaining free memory for the warning (e.g. warn if free memory is less than 1 GiB), but this is just a heuristic, we assumed that on systems with a lot of memory, other processes are also running, while memory constrained ones run only Core.

If total RAM < 2 GiB, cap is `DEFAULT_DB_CACHE` (`450 MiB`), otherwise it's 75% of total RAM.
The threshold is chosen to be close to values commonly used in [raspiblitz](https://github.com/raspiblitz/raspiblitz/blob/dev/home.admin/_provision.setup.sh#L98-L115) for common setups:

| Total RAM | `dbcache` (MiB) | raspiblitz % | proposed cap (MiB) |
|----------:|----------------:|-------------:|-------------------:|
|     1 GiB |             512 |        50.0% |               450* |
|     2 GiB |            1536 |        75.0% |               1536 |
|     4 GiB |            2560 |        62.5% |               3072 |
|     8 GiB |            4096 |        50.0% |               6144 |
|    16 GiB |            4096 |        25.0% |              12288 |
|    32 GiB |            4096 |        12.5% |              24576 |

[Umbrel issues](https://github.com/getumbrel/umbrel-os/issues/64#issuecomment-663816367) also mention 75% being the upper limit.

Starting `bitcoind` on an 8 GiB rpi4b with a dbcache of 7 GiB:
> ./build/bin/bitcoind -dbcache=7000

warns now as follows:
```
2025-09-07T17:24:29Z [warning] A 7000 MiB dbcache may be too large for a system memory of only 7800 MiB.
2025-09-07T17:24:29Z Cache configuration:
2025-09-07T17:24:29Z * Using 2.0 MiB for block index database
2025-09-07T17:24:29Z * Using 8.0 MiB for chain state database
2025-09-07T17:24:29Z * Using 6990.0 MiB for in-memory UTXO set (plus up to 286.1 MiB of unused mempool space)
```

Besides the [godbolt](https://godbolt.org/z/EPsaE3xTj) reproducers for the new total memory method, we also tested the warnings manually on:
- [x] Apple M4 Max, macOS 15.6.1
- [x] Intel Core i9-9900K, Ubuntu 24.04.2 LTS
- [x] Raspberry Pi 4 Model B, Armbian Linux 6.12.22-current-bcm2711
- [x] Intel Xeon x64, Windows 11 Home Version 24H2, OS Build 26100.4351

Co-authored-by: stickies-v <stickies-v@protonmail.com>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-authored-by: w0xlt <woltx@protonmail.com>

TheCharlatan commented:
Concept ACK

w0xlt reviewed (commented):
Approach ACK

I’d suggest:
. Logging the `else` condition as well, for consistency.
. Showing the percentage of RAM being used.

Example:

```diff
@@ -1736,9 +1736,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     const auto [index_cache_sizes, kernel_cache_sizes] = CalculateCacheSizes(args, g_enabled_filter_types.size());
     if (auto total_ram{GetTotalRAM()}) {
         size_t cap{(*total_ram < 2048_MiB) ? DEFAULT_KERNEL_CACHE : (*total_ram / 100) * 75};
+        size_t actual_percentage = (kernel_cache_sizes.coins * 100) / *total_ram;
         if (kernel_cache_sizes.coins > cap) {
-            LogWarning("A %zu MiB dbcache may be too large for a system memory of only %zu MiB.", kernel_cache_sizes.coins >> 20, *total_ram >> 20);
-        }
+            LogWarning("A %zu MiB dbcache may be too large for a system memory of only %zu MiB (%zu%% of system memory).",
+                    kernel_cache_sizes.coins >> 20, *total_ram >> 20, actual_percentage);
+        } else {
+            LogInfo("Total system memory is %zu MiB; setting dbcache to %zu MiB (%zu%% of system memory)",
+                    *total_ram >> 20, kernel_cache_sizes.coins >> 20, actual_percentage);
+    }
     }

     LogInfo("Cache configuration:");
```

stickies-v reviewed (commented):
Concept ACK

hodlinator reviewed (commented):
Concept ACK

Confirmed to work on Windows:
```
$ ./build/bin/Release/bitcoind.exe -regtest -dbcache=64534
2025-09-12T22:14:09Z Bitcoin Core version v29.99.0-080a21cb127f-dirty (release build)
...
2025-09-12T22:14:09Z [warning] A 64524 MiB dbcache may be too large for a system memory of only 65413 MiB.
```
Edit: Intel Xeon x64, Windows 11 Home Version 24H2, OS Build 26100.4351
Edit2: The above was a Windows -> Windows build, going to try Guix cross build also, stay posted.
Edit3: Guix Linux -> Windows cross build also works:
```
C:\Program Files\Bitcoin>bitcoin node -regtest -dbcache=65000
C:\Program Files\Bitcoin>2025-09-13T18:53:35Z Bitcoin Core version v29.99.0-g080a21cb127f901d59b8478404db513a0ce0ed07 (release build)
...
2025-09-13T18:53:35Z [warning] A 64990 MiB dbcache may be too large for a system memory of only 65413 MiB.
```

l0rinc committed:
system: add helper for fetching total system memory

Added a minimal system helper to query total physical RAM on [Linux/macOS/Windows](https://stackoverflow.com/a/2513561) (on other platforms we just return an empty optional).

The added test checks if the value is roughly correct by checking if the CI platforms are returning any value and if the value is at least 1 GiB and not more than 10 TiB.

The max value is only validated on 64 bits, since it's not unreasonable for 32 bits to have max memory, but on 64 bits it's likely an error.

https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-memorystatusex
> ullTotalPhys The amount of actual physical memory, in bytes.

https://man7.org/linux/man-pages/man3/sysconf.3.html:
> _SC_PHYS_PAGES The number of pages of physical memory. Note that it is possible for the product of this value and the value of _SC_PAGESIZE to overflow.
> _SC_PAGESIZE Size of a page in bytes. Must not be less than 1.

See https://godbolt.org/z/ec81Tjvrj for further details

l0rinc commented on src/init.cpp:
```diff
@@ -1734,6 +1734,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    if (auto total_ram{GetTotalRAM()}) {
+        size_t cap{(*total_ram < 2048_MiB) ? DEFAULT_KERNEL_CACHE : (*total_ram / 100) * 75};
+        if (kernel_cache_sizes.coins > cap) {
+            LogWarning("A %zu MiB dbcache may be too large for a system memory of only %zu MiB.", kernel_cache_sizes.coins >> 20, *total_ram >> 20);
```
Added this in a separate commit (I'm not on my phone anymore :p), reviewers can check the commit message for instructions for how to reproduce it

l0rinc reviewed (commented):
Thanks a lot for the comments and reproducers, added a @hodlinator and @stickies-v as coauthors.

I have split the PR into 3 commits, explained the descisions in the messages (cc: @willcl-ark), added unit tests for whatever I could extract, extracted the warning message to a method and exposed the warning for RPC (thanks @stickies-v).

hodlinator commented on src/common/system.cpp:
```diff
@@ -105,6 +111,21 @@ int GetNumCores()
+{
+#ifdef WIN32
+    if (MEMORYSTATUSEX m{}; (m.dwLength = sizeof(m), GlobalMemoryStatusEx(&m))) {
+        return size_t(std::clamp<uint64_t>(m.ullTotalPhys, 0, SIZE_MAX));
```
Think we should rely on the `NOMINMAX` define you found in *bitcoin/CMakeLists.txt*, also prefer `numeric_limits`. `std::clamp<uint64_t>(` with the lower bound of `0` wasn't doing much so agree with getting rid of that.

~0 on switching from separate `if`-blocks to mutating `mem`-variable.

hodlinator commented on src/test/caches_tests.cpp:
```diff
@@ -0,0 +1,40 @@
+
+BOOST_AUTO_TEST_SUITE(caches_tests)
+
+BOOST_AUTO_TEST_CASE(oversized_dbcache_warning)
+{
+    // memory restricted setup - cap is DEFAULT_DB_CACHE (450 MiB)
+    BOOST_CHECK(!ShouldWarnOversizedDbCache(/*dbcache=*/4_MiB, /*total_ram=*/1024_MiB));    // Under cap
+    BOOST_CHECK( ShouldWarnOversizedDbCache(/*dbcache=*/512_MiB, /*total_ram=*/1024_MiB));  // At cap
+    BOOST_CHECK( ShouldWarnOversizedDbCache(/*dbcache=*/1500_MiB, /*total_ram=*/1024_MiB)); // Over cap
+
+    // 2 GiB RAM - cap is 75%
+    BOOST_CHECK(!ShouldWarnOversizedDbCache(/*dbcache=*/1500_MiB, /*total_ram=*/2048_MiB)); // Under cap
+    BOOST_CHECK( ShouldWarnOversizedDbCache(/*dbcache=*/1600_MiB, /*total_ram=*/2048_MiB)); // Over cap
+
+    if constexpr (SIZE_MAX == UINT64_MAX) {
+        // 4 GiB RAM - cap is 75%
+        BOOST_CHECK(!ShouldWarnOversizedDbCache(/*dbcache=*/2500_MiB, /*total_ram=*/4096_MiB)); // Under cap
+        BOOST_CHECK( ShouldWarnOversizedDbCache(/*dbcache=*/3500_MiB, /*total_ram=*/4096_MiB)); // Over cap
+
+        // 8 GiB RAM - cap is 75%
+        BOOST_CHECK(!ShouldWarnOversizedDbCache(/*dbcache=*/6000_MiB, /*total_ram=*/8192_MiB)); // Under cap
+        BOOST_CHECK( ShouldWarnOversizedDbCache(/*dbcache=*/7000_MiB, /*total_ram=*/8192_MiB)); // Over cap
+
+        // 16 GiB RAM - cap is 75%
+        BOOST_CHECK(!ShouldWarnOversizedDbCache(/*dbcache=*/10'000_MiB, /*total_ram=*/16384_MiB)); // Under cap
+        BOOST_CHECK( ShouldWarnOversizedDbCache(/*dbcache=*/15'000_MiB, /*total_ram=*/16384_MiB)); // Over cap
+
+        // 32 GiB RAM - cap is 75%
+        BOOST_CHECK(!ShouldWarnOversizedDbCache(/*dbcache=*/20'000_MiB, /*total_ram=*/32768_MiB)); // Under cap
+        BOOST_CHECK( ShouldWarnOversizedDbCache(/*dbcache=*/30'000_MiB, /*total_ram=*/32768_MiB)); // Over cap
+    }
+}
```
(nit: as stated previously I prefer making simple functions like `ShouldWarnOversizedDbCache` `constexpr` and using `static_assert` instead of runtime checks, https://github.com/bitcoin/bitcoin/pull/31645#discussion_r2282516167).

hodlinator commented on src/node/caches.cpp:
```diff
@@ -23,16 +27,22 @@ static constexpr size_t MAX_FILTER_INDEX_CACHE{1024_MiB};
 namespace node {
-CacheSizes CalculateCacheSizes(const ArgsManager& args, size_t n_indexes)
+size_t CalculateDbCacheBytes(const ArgsManager& args)
 {
-    // Convert -dbcache from MiB units to bytes. The total cache is floored by MIN_DB_CACHE and capped by max size_t value.
-    size_t total_cache{DEFAULT_DB_CACHE};
+    size_t total_cache = {DEFAULT_DB_CACHE};
```
Stray addition of ` = ` in last commit.

Suggestion: Could remove `total_cache` completely (the name is more fitting for the original copy which is still initialized by the result of this function anyway):
```diff
--- a/src/node/caches.cpp
+++ b/src/node/caches.cpp
@@ -29,14 +29,14 @@ static constexpr size_t MAX_32BIT_DBCACHE{1024_MiB};
 namespace node {
 size_t CalculateDbCacheBytes(const ArgsManager& args)
 {
-    size_t total_cache = {DEFAULT_DB_CACHE};
     if (std::optional<int64_t> db_cache = args.GetIntArg("-dbcache")) {
         if (*db_cache < 0) db_cache = 0;
         const uint64_t db_cache_bytes{SaturatingLeftShift<uint64_t>(*db_cache, 20)};
         constexpr auto max_db_cache{sizeof(void*) == 4 ? MAX_32BIT_DBCACHE : std::numeric_limits<size_t>::max()};
-        total_cache = std::max<size_t>(MIN_DB_CACHE, std::min<uint64_t>(db_cache_bytes, max_db_cache));
+        return std::max<size_t>(MIN_DB_CACHE, std::min<uint64_t>(db_cache_bytes, max_db_cache));
+    } else {
+        return DEFAULT_DB_CACHE;
     }
-    return total_cache;
 }
 
 CacheSizes CalculateCacheSizes(const ArgsManager& args, size_t n_indexes)
```

hodlinator commented on src/node/caches.cpp:
```diff
@@ -44,4 +54,17 @@ CacheSizes CalculateCacheSizes(const ArgsManager& args, size_t n_indexes)
+            const auto msg{tfm::format(_("A %zu MiB dbcache may be too large for a system memory of only %zu MiB."),
+                db_cache_bytes >> 20, *total_ram_bytes >> 20)};
+            LogWarning("%s", msg.original);
+            Assert(warnings.Set(kernel::Warning::OVERSIZED_DBCACHE, msg));
```
Might be overly harsh with `Assert()`? Could use `Assume()`, and/or maybe add a comment why this is worth guarding against. Noticed the original suggestion didn't even check the return value (https://github.com/bitcoin/bitcoin/pull/33333#discussion_r2344204033).

hodlinator reviewed (commented):
Reviewed 9f69c646d7eb8790ec9fd6d1d5607f554cc2f4ff

l0rinc commented on src/node/caches.cpp:
```diff
@@ -23,16 +27,22 @@ static constexpr size_t MAX_FILTER_INDEX_CACHE{1024_MiB};
+
+CacheSizes CalculateCacheSizes(const ArgsManager& args, size_t n_indexes)
+{
+    // Convert -dbcache from MiB units to bytes. The total cache is floored by MIN_DB_CACHE and capped by max size_t value.
```
In that case I'd just delete it since the code already states all of this

l0rinc reviewed (commented):
Thanks for the reviews, keep them coming :)

achow101 commented:
I think it would be better to use an `InitWarning` rather than just logging since people are unlikely notice

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/33333/commits/13fa0d0a433722f294854001b0561c079db12dbc

BSD operating system support can be a follow-up.

hodlinator reviewed (approved):
ACK 13fa0d0a433722f294854001b0561c079db12dbc

Very helpful to warn users against unintentionally making their system start swapping insanely.

danielabrozzoni reviewed (commented):
tACK 13fa0d0a433722f294854001b0561c079db12dbc

I tested locally and reproduced the warning:

```
❯ ./build/bin/bitcoind -dbcache=64000
...
2025-09-16T13:57:22Z [warning] A 64000 MiB dbcache may be too large for a system memory of only 64219 MiB.
Warning: A 64000 MiB dbcache may be too large for a system memory of only 64219 MiB.

❯ ./build/bin/bitcoin-cli getblockchaininfo
{
  [...]
  "warnings": [
    "A 64000 MiB dbcache may be too large for a system memory of only 64219 MiB.",
    "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
  ]
}
```      

I left a nit, feel free to ignore if you think it's unnecessary or if you don't need to retouch

l0rinc commented on src/test/system_tests.cpp:
```diff
@@ -16,6 +18,15 @@
+{
+    BOOST_CHECK_GE(GetTotalRAM(), 1000_MiB);
+
+    if constexpr (SIZE_MAX == UINT64_MAX) {
```
I have explained it in the commit message
> The added test checks if the value is roughly correct by checking if the CI platforms are returning any value and if the value is at least 1 GiB and not more than 1 TiB.

The max value is only validated on 64 bits, since it's not unreasonable for 32 bits to have max memory, but on 64 bits it's likely an error.
Do you think it needs more explanation than that?

l0rinc replied:
Added comment and updated commit message

l0rinc commented:
Thanks for the feedback, removed the last commit.

l0rinc commented on src/test/system_tests.cpp:
```diff
@@ -16,6 +18,15 @@
+    BOOST_CHECK_GE(GetTotalRAM(), 1000_MiB);
+
+    if constexpr (SIZE_MAX == UINT64_MAX) {
+        BOOST_CHECK_LT(GetTotalRAM(), 1'000'000_MiB);
```
that's what the previous check does on line 23, this validates that it's in a reasonable range, e.g. we're not returning number of bits or something

achow101 replied:
>1 TB of RAM is entirely feasible on high end workstations and servers, I think we could add a couple more 0's to this check.

l0rinc replied:
Done

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/33333/commits/168360f4ae47cbfdb30a2cc4704435bc67e12f16

hodlinator reviewed (approved):
re-ACK 168360f4ae47cbfdb30a2cc4704435bc67e12f16

danielabrozzoni reviewed (commented):
reACK 168360f4ae47cbfdb30a2cc4704435bc67e12f16

hebasto commented on src/common/system.cpp:
```diff
@@ -105,6 +111,17 @@ int GetNumCores()
     return std::thread::hardware_concurrency();
 }
 
+std::optional<size_t> GetTotalRAM()
+{
+    auto clamp{[](uint64_t v) { return size_t(std::min(v, uint64_t{std::numeric_limits<size_t>::max()})); }};
+#ifdef WIN32
+    if (MEMORYSTATUSEX m{}; (m.dwLength = sizeof(m), GlobalMemoryStatusEx(&m))) return clamp(m.ullTotalPhys);
+#elif defined(__linux__) || defined(__APPLE__)
+    if (long p{sysconf(_SC_PHYS_PAGES)}, s{sysconf(_SC_PAGESIZE)}; p > 0 && s > 0) return clamp(1ULL * p * s);
+#endif
+    return std::nullopt;
+}
+
```
This should work:
```suggestion
std::optional<size_t> GetTotalRAM()
{
    auto clamp{[](uint64_t v) { return size_t(std::min(v, uint64_t{std::numeric_limits<size_t>::max()})); }};
#ifdef WIN32
    if (MEMORYSTATUSEX m{}; (m.dwLength = sizeof(m), GlobalMemoryStatusEx(&m))) return clamp(m.ullTotalPhys);
#else
    if (long p{sysconf(_SC_PHYS_PAGES)}, s{sysconf(_SC_PAGESIZE)}; p > 0 && s > 0) return clamp(1ULL * p * s);
#endif
    return std::nullopt;
}

```

See CI outputs for:
- FreeBSD: https://github.com/hebasto/bitcoin-core-nightly/actions/runs/17856236916
- OpenBSD: https://github.com/hebasto/bitcoin-core-nightly/actions/runs/17856241624
- NetBSD: https://github.com/hebasto/bitcoin-core-nightly/actions/runs/17856250035
- OmniOS: https://github.com/hebasto/bitcoin-core-nightly/actions/runs/17856253573

vasild replied:
I came here for the same reason ;-) Here is the patch I devised:

```diff
 std::optional<size_t> GetTotalRAM()
 {
-    auto clamp{[](uint64_t v) { return size_t(std::min(v, uint64_t{std::numeric_limits<size_t>::max()})); }};
+    [[maybe_unused]] auto clamp{[](uint64_t v) { return size_t(std::min(v, uint64_t{std::numeric_limits<size_t>::max()})); }};
 #ifdef WIN32
     if (MEMORYSTATUSEX m{}; (m.dwLength = sizeof(m), GlobalMemoryStatusEx(&m))) return clamp(m.ullTotalPhys);
-#elif defined(__linux__) || defined(__APPLE__)
+#elif defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__)
     if (long p{sysconf(_SC_PHYS_PAGES)}, s{sysconf(_SC_PAGESIZE)}; p > 0 && s > 0) return clamp(1ULL * p * s);
 #endif
     return std::nullopt;
 }
```

I did not go with `#else` since I am not sure that will compile and work on all possible platforms that are not Windows.

vasild replied:
PRed in https://github.com/bitcoin/bitcoin/pull/33435

fanquake commented:
Backported to 30.x in #33424.
